### PR TITLE
fix/feat: New cleanup implementation

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-416

:pray: Please, check it commit by commit

**Changes:**
- Replaced old implementation of the `storage/cleanup`.
  - It deletes metadata of old files/slices from the DB.

-----------

### CODE REVIEW

File: internal/pkg/service/stream/storage/test/testconfig/testconfig.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/volume"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/volume/assignment"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/volume/disksync"
+)
+
+func StorageConfigPatch() configpatch.PatchKVs {
+	return configpatch.MustDumpPatch(
+		config.Config{},
+		config.Patch{
+			Storage: &storage.ConfigPatch{
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/test/testconfig/testconfig.go:23]: The function generates `KV` pairs.

File: internal/pkg/service/stream/cleanup/node.go [hunk 2]
```diff
@@ -1,115 +0,0 @@
...
-}
-
-func (n *Node) Check(ctx context.Context) error {
-	tasksCount := 0
-	err := n.schema.
-		Configs().
-		Receivers().
-		GetAll().
-		Do(ctx, n.client).
-		ForEachValue(func(v model.ReceiverBase, header *iterator.Header) error {
```
Comment (michaljurecko) [inline internal/pkg/service/stream/cleanup/node.go:69]: Removed old cleanup implementation, it was iterating `receivers/sources` and a task was created for each source.

Not all files are iterated directly (there is an interator), so it should be faster and simpler.

File: internal/pkg/service/common/dependencies/distribution.go [hunk 3]
```diff
@@ -18,15 +18,15 @@ type distributionScopeDeps interface {
 	TaskScope
 }
 
-func NewDistributionScope(ctx context.Context, nodeID string, d distributionScopeDeps) (DistributionScope, error) {
-	return newDistributionScope(ctx, nodeID, d)
+func NewDistributionScope(ctx context.Context, nodeID string, cfg distributionPkg.Config, d distributionScopeDeps) (DistributionScope, error) {
+	return newDistributionScope(ctx, nodeID, cfg, d)
 }
```
Comment (michaljurecko) [inline internal/pkg/service/common/dependencies/distribution.go:24]: Distribution config can be now modfied.

File: internal/pkg/service/stream/storage/repository/file_repo.go [hunk 4]
```diff
@@ -454,10 +454,8 @@ func (r *FileRepository) rotateSink(ctx context.Context, c rotateSinkContext) (*
 	if c.NewFileResource != nil {
 		// Apply configuration patch from the sink to the global config
 		cfg := r.config
-		if c.Sink.Table.Config.Storage.Level != nil {
-			if err := configpatch.Apply(&cfg, c.Sink.Table.Config.Storage.Level); err != nil {
-				return nil, err
-			}
+		if err := configpatch.ApplyKVs(&cfg, &level.ConfigPatch{}, c.Sink.Config.In("storage.level")); err != nil {
+			return nil, err
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/repository/file_repo.go:458]: Updatedm `c.Sink.Config` now contains KV pairs.

File: internal/pkg/service/stream/dependencies/dependencies.go [hunk 5]
```diff
@@ -69,6 +69,8 @@ type DefinitionScope interface {
 
 type TableSinkScope interface {
 	DefinitionScope
+	dependencies.DistributionScope
+	dependencies.DistributedLockScope
 	StatisticsRepository() *statsRepo.Repository
 	StatisticsL1Cache() *cache.L1
```
Comment (michaljurecko) [inline internal/pkg/service/stream/dependencies/dependencies.go:75]: Added scopes needed by the `storage/cleanup`.

File: api/stream/design.go [hunk 6]
```diff
@@ -703,8 +703,8 @@ var TableID = Type("TableID", String, func() {
 })
 
 var TableColumns = Type("TableColumns", ArrayOf(TableColumn), func() {
-	minLength := cast.ToInt(fieldValidationRule(definition.TableSinkMapping{}, "Columns", "min"))
-	maxLength := cast.ToInt(fieldValidationRule(definition.TableSinkMapping{}, "Columns", "max"))
+	minLength := cast.ToInt(fieldValidationRule(table.Mapping{}, "Columns", "min"))
+	maxLength := cast.ToInt(fieldValidationRule(table.Mapping{}, "Columns", "max"))
 	Description(fmt.Sprintf("List of export column mappings. An export may have a maximum of %d columns.", maxLength))
```
Comment (michaljurecko) [inline api/stream/design.go:708]: See comment bellow.

File: internal/pkg/service/stream/storage/test/file.go [hunk 7]
```diff
@@ -80,9 +80,13 @@ func NewFileOpenedAt(openedAtStr string) model.File {
...
-			TableID:    keboola.MustParseTableID("in.bucket.table"),
-			StorageJob: nil,
+		TargetStorage: target.Target{
+			Table: target.Table{
+				Keboola: target.KeboolaTable{
+					TableID:    keboola.MustParseTableID("in.bucket.table"),
+					StorageJob: nil,
+				},
+			},
 		},
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/test/file.go:90]: Improved forward compatibility.

Now, we are importing data to the Keboola table, using a Storage Job.
There are some limits, it is safe to run at most 1 Storage Job each `30s`.

In the future, we will probably write directly to the table. So there will be several configuration options.

File: internal/pkg/service/stream/config/config.go [hunk 8]
```diff
@@ -21,18 +22,25 @@ const (
...
+	DebugLog        bool                `configKey:"debugLog"  configUsage:"Enable logging at DEBUG level."`
+	DebugHTTPClient bool                `configKey:"debugHTTPClient" configUsage:"Log HTTP client requests and responses as debug messages."`
+	CPUProfFilePath string              `configKey:"cpuProfilePath" configUsage:"Path where CPU profile is saved."`
+	NodeID          string              `configKey:"nodeID" configUsage:"Unique ID of the node in the cluster." validate:"required"`
+	StorageAPIHost  string              `configKey:"storageAPIHost" configUsage:"Storage API host." validate:"required,hostname"`
+	Datadog         datadog.Config      `configKey:"datadog"`
+	Etcd            etcdclient.Config   `configKey:"etcd"`
+	Metrics         prometheus.Config   `configKey:"metrics"`
+	API             API                 `configKey:"api"`
+	Distribution    distribution.Config `configKey:"distribution"`
```
Comment (michaljurecko) [inline internal/pkg/service/stream/config/config.go:34]: Added `Distribution` config.

File: internal/pkg/service/stream/storage/config.go [hunk 9]
```diff
@@ -1,26 +1,26 @@
...
 	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/statistics"
 )
 
-// Config contains default configuration for the storage.
-// Default settings can be customized in definition.TableSink.
+// Config contains global configuration for the storage.
 type Config struct {
 	Statistics statistics.Config `configKey:"statistics"`
+	Cleanup    cleanup.Config    `configKey:"cleanup"`
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/config.go:12]: Added cleanup config.

File: internal/pkg/service/stream/definition/sink_base.go [hunk 10]
```diff
@@ -11,10 +12,14 @@ type Sink struct {
...
 	Switchable
 	SoftDeletable
-	Type        SinkType   `json:"type" validate:"required,oneof=table"`
-	Name        string     `json:"name" validate:"required,min=1,max=40"`
-	Description string     `json:"description,omitempty" validate:"max=4096"`
-	Table       *TableSink `json:"table" validate:"required_if=Type table"`
+	Type        SinkType             `json:"type" validate:"required,oneof=table"`
+	Name        string               `json:"name" validate:"required,min=1,max=40"`
+	Description string               `json:"description,omitempty" validate:"max=4096"`
+	Config      configpatch.PatchKVs `json:"config"` // see stream/config/config.Patch
```
Comment (michaljurecko) [inline internal/pkg/service/stream/definition/sink_base.go:18]: Sink now contains configuration patch as KV pairs. 

So we can modify anything provided by the `config.Patch`, but it is not directry coupled with the `config` package.

File: internal/pkg/service/stream/config/config_test.go [hunk 11]
```diff
@@ -153,4 +187,302 @@ storage:
...
+	cfg.StorageAPIHost = "connection.keboola.local"
+	cfg.Etcd.Endpoint = "test-etcd"
+	cfg.Etcd.Namespace = "test-namespace"
+	require.NoError(t, validator.New().Validate(context.Background(), cfg))
+}
+
+func TestTableSinkConfigPatch_ToKVs(t *testing.T) {
+	t.Parallel()
+
+	kvs, err := configpatch.DumpAll(
```
Comment (michaljurecko) [inline internal/pkg/service/stream/config/config_test.go:202]: Moved from the `tablesink/config_test.go`.

File: internal/pkg/service/stream/storage/cleanup/node.go [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+		}
+	}()
+
+	return nil
+}
+
+// cleanFiles iterates all files and deletes the expired ones.
+func (n *Node) cleanFiles(ctx context.Context) (err error) {
+	ctx, span := n.telemetry.Tracer().Start(ctx, "keboola.go.stream.model.cleanup.node.cleanFiles")
+	defer span.End(&err)
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/cleanup/node.go:107]: The new cleanup implementation.

File: internal/pkg/service/stream/storage/cleanup/node_test.go [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/test/testconfig"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/etcdhelper"
+)
+
+func TestNode(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+
+	clk := clock.NewMock()
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/cleanup/node_test.go:32]: The hardest part - the test.

File: internal/pkg/service/stream/storage/cleanup/node.go [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+
+	// Acquire lock
+	mutex := n.locks.NewMutex(file.FileKey.String())
+	if err = mutex.TryLock(ctx); err != nil {
+		return err, false
+	}
+	defer func() {
+		if err := mutex.Unlock(ctx); err != nil {
+			n.logger.Errorf(ctx, "cannot unlock the lock: %s", err)
+		}
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/cleanup/node.go:174]: Distributed lock.

File: internal/pkg/service/stream/storage/cleanup/node.go [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+	}()
+
+	// Delete the file
+	if err = n.storage.File().Delete(file.FileKey).RequireLock(mutex).Do(ctx).Err(); err != nil {
```
Comment (michaljurecko) [inline internal/pkg/service/stream/storage/cleanup/node.go:178]: `RequireLock(mutex)` - distributed lock - token fencing.

https://medium.com/@lalitadithya/everything-ive-learnt-about-distributed-locking-so-far-1f1569e6df5

`# Fenced locks`

image

File: internal/pkg/service/stream/storage/cleanup/node.go [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+	}
+
+	ctx := context.Background()
+	if !n.config.Enabled {
+		n.logger.Info(ctx, "storage cleanup is disabled")
+		return nil
+	}
+
+	// Graceful shutdown
+	ctx, cancel := context.WithCancel(ctx) // nolint: contextcheck
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/cleanup/node.go:70]: This is different case, and it is `context.Background()`, isn't it?

image

[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/cleanup/node.go]: I removed unused `nolint` comment: https://github.com/keboola/keboola-as-code/pull/1600/commits/43c1c1591051ee15752025ade11ac412096e11bc

File: internal/pkg/service/stream/storage/cleanup/node.go [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+func (n *Node) cleanFiles(ctx context.Context) (err error) {
+	ctx, span := n.telemetry.Tracer().Start(ctx, "keboola.go.stream.model.cleanup.node.cleanFiles")
+	defer span.End(&err)
+
+	// Measure count of deleted files
+	count := atomic.NewInt64(0)
+	defer func() {
+		cnt := count.Load()
+		span.SetAttributes(attribute.Int64("deletedFilesCount", cnt))
+		n.logger.With(attribute.Int64("count", cnt)).Infof(ctx, `deleted "%d" files`, cnt)
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/stream/storage/cleanup/node.go:114]: Yes, work well, but is may be improved, ... it was already discussed in the past.

image

https://github.com/keboola/keboola-as-code/pull/1606

**Comment by michaljurecko** [comment]
ping @hosekpeter :pray:

**Review by hosekpeter — APPROVED** [review]

**Pull request merged**